File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <vector>

using namespace std;

long long solution(int n, vector<int> times) {

    long long lo = 0;
    long long hi = times.size() * 1000000000;
    
    // mid 값은 시간
    // n 보다 크면 시간을 줄이면 된다.(hi에서 값이 나올듯.)
    while(lo < hi) {
        
        long long mid = (lo + hi) / 2;
        long long cnt = 0;
        for(int time : times) {
            cnt += mid / time;
        }
        
        if(cnt>= n)
            hi = mid;
        else
            lo = mid + 1;
        
    }
    
    return hi;
}

File: 프로그래머스/3/43238. 입국심사/입국심사.java
class Solution {
    public long solution(int n, int[] times) {
        long answer = (long)Math.pow(10,9) * (long) Math.pow(10, 9);
        
        long lo = 1, hi = answer;
        long ans = 0;
        
        while(lo <= hi){
            long mid = (hi -lo)/2 + lo;
            long cnt = 0;
            for(int t : times){
                cnt += mid / t;
            }
            
            if(cnt >= n){
                ans = mid;
                hi = mid-1;
            }else{
                lo = mid+1;
            }
        }
        
        return ans;
    }
}